doc/erofs.md

```diff
@@ -0,0 +1,431 @@
+*   `uid`, `gid`: those are fairly obvious, I guess
+*   `mtime`, `mtime_nsec`: those too
+*   `nlink`: try to set this correctly: some things might get upset if it's not
+    right.  For non-directories, that's the number of hardlinks (ie: 1 for
+    non-hardlinked files).  For directories, that's 2 plus the number of
+    subdirectories.
+
+Directly following the inode header is the extended attribute header (if
+`xattr_icount` is non-zero).  Then comes any inline data (as per the "data
+layout" section).
+
+### Extended attributes
+
+If the `xattr_icount` field in the inode header is set to 0 then this section
+is skipped entirely.  Otherwise we write out the inode xattr header (`struct
+erofs_xattr_ibody_header`).  This has:
+*   `name_filter` (`u32`): a bloom filter for which xattrs are present.  This
+    needs its own section.
+*   `shared_count` (`u8`): the number of shared xattrs
+*   some reserved bytes to pad things up to 12
+
+Immediately following the header come the shared xattr references.  They're in
+the format mentioned in the "Shared xattrs" section above, simply encoded as
+little-endian u32s.  So: the first `4 * shared_count` bytes after the header
+are those.
+
+Then the inline xattrs are next.  Those are stored in the format mentioned in
+the "On-disk format" sub-section in the "Extended attributes" section.  They're
+just written here one after another, with padding added so that each header is
+4-aligned.  There is also padding after the last one, which is important if
+inline data is to follow (as per the "data layout" section).
+
+#### About `xattr_icount`
+
+So, if there's no xattrs then this is zero.
+
+Otherwise this is basically the size of the extended attributes area divided by
+4, with the exception that the 12-byte header counts for only 4 bytes.  Put
+another way: you remove the size of the header, divide by 4, then add 1 back
+again.
+
+A value of 1 would be pretty suspicious, since that would indicate the presence
+of a header, but no xattrs (shared or inline), and in that case normally we'd
+omit the header.
+
+The kernel basically uses this to know how many bytes it needs to skip over
+before it can find the inline file data.  It will remove the 1, multiply by 4,
+then add 12 (the header).  See `erofs_xattr_ibody_size()`.
+
+#### About `name_filter`
+
+This is a 32-bit bloom filter used to quickly determine if a given xattr is not present.
+
+The hash algorithm is xxh32.  The thing that gets hashed is not the name, but
+the "suffix" that's left after removing the prefix.  The seed is
+`EROFS_XATTR_FILTER_SEED` plus the prefix index.  The lower 5 bits of the hash
+value (0..31) are used to determine which bit is used.
+
+For some reason a bit value of 1 here indicates the absence of a particular
+xattr, which is opposite to the usual arrangement.  You'd think it was for
+compatibility, but the filter is only engaged if the feature bit is present in
+the superblock.
```

Comment:
that is due to incremental builds too, because some on-disk inodes could be previously generated and we don't need to touch them to do an incremental build

Reply:
Right!  Makes sense, thanks!